### PR TITLE
MTDSA-22029 - TYS schemas V5

### PR DESCRIPTION
**Pull request by jeremystone**
### Notes:
* Created full v5 classes (from v3 and v4 where reqd) 

* Used _endpoint name-first_ package structure with child packages `controllers`, `services`, `models` etc. This may need changing if we go with something different, but it is much easier to find things than the old structure.

* Have gone with the class naming convention as discussed (namely only prefix Def1_ for the top-level request data, request body and response classes). However, for list (hateoas wrapping) some classes deeper in the data structures also required schema-independent traits with Def1-prefixed implementations. 

* One child package is new: `schema`. This contains an enumeration (via sealed trait) of schema value Def1, Def2 etc for the endpoint that represent the different schema definitions. The purpose of this is two-fold:
    * To allow the schema value to be passed along as part of the request data to be used. Currently this is only used in the connectors (but is not reqd for empty/ignored downstream responses) to determine the type of JSON `Reads` the connector needs to use. Connectors simply use `import schema._` after the `import request._` to bring it into scope and change the type parameters on the `DownstreamUri` to `DownstreamResp`.
    * The other purpose is that it can then be used explicitly in the validator factory to select the appropriate validator.

* For each endpoint the schema trait companion object has `schemaFor` method(s) to get the schema value based on the tax year. The signature of this method depends on where the tax year comes from: typically a query or path param, but BSAS trigger has it in the request body. Because this is done prior to validation, a default Def1 schema is used if there is any problem with getting the tax year. This _seems_ reasonable but we might want to give this further thought. There is certainly some scope for cleaner ways to map tax year ranges to schema values.

### CODE REVIEW

File: app/v5/controllers/validators/resolvers/ResolveOnlyOneJsonProperty.scala [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+package v5.controllers.validators.resolvers
+
+import cats.data.Validated
+import cats.data.Validated.{Invalid, Valid}
+import play.api.libs.json.JsValue
+import shared.models.errors.MtdError
+import v5.models.errors.RuleBothPropertiesSuppliedError
+
+class ResolveOnlyOneJsonProperty(fieldOneName: String, fieldTwoName: String) {
```
Comment (mattstephens-hmrc) [inline app/v5/controllers/validators/resolvers/ResolveOnlyOneJsonProperty.scala:25]: This one seems fairly gen-purpose, could go in the shared resolvers?

Comment (jeremystone) [reply app/v5/controllers/validators/resolvers/ResolveOnlyOneJsonProperty.scala]: Moved to shared and renamed `ResolveExclusiveJsonProperty`. Also:

-  simplified its usage so no need to use `productR`. This reverts behaviour for edge case where both properties are supplied _and_ they also happen to be empty to what it used to be (see [here](https://github.com/hmrc/self-assessment-bsas-api/blob/0433b3fc587749c281f62db7dfe824b7c457a9d9/test/v3/controllers/requestParsers/validators/SubmitUkPropertyBsasValidatorSpec.scala#L231)).
- Moved resolver instance creation to the validator companions so that they can be shared between requests.

**Review by abigail-houston-hmrc — APPROVED** [review]
